yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtil.scala

```diff
@@ -82,6 +103,244 @@ class YarnSparkHadoopUtil extends SparkHadoopUtil {
+                return
+            }
+            scheduleRenewal(this)
+          }
+        }
+      // If this is an AM restart, it is possible that the original tokens have expired, which
+      // means we need to login immediately to get new tokens.
+      if (getRenewalInterval == 0) writeNewTokensToHDFS(principal, keytab)
+      // Schedule update of credentials
+      scheduleRenewal(driverTokenRenewerRunnable)
+    }
+  }
+
+  private def writeNewTokensToHDFS(principal: String, keytab: String): Unit = {
+    if (!loggedInViaKeytab) {
+      // Keytab is copied by YARN to the working directory of the AM, so full path is
+      // not needed.
+      logInfo(s"Attempting to login to KDC using principal: $principal")
+      loggedInUGI = UserGroupInformation.loginUserFromKeytabAndReturnUGI(
+        principal, keytab)
+      logInfo("Successfully logged into KDC.")
+      loggedInViaKeytab = true
+      // Not exactly sure when HDFS re-logs in, be safe and do it ourselves.
+      // Periodically check and relogin this keytab. The UGI will take care of not relogging in
+      // if it is not necessary to relogin.
+      val reloginRunnable = new Runnable {
+        override def run(): Unit = {
+          try {
+            loggedInUGI.checkTGTAndReloginFromKeytab()
+          } catch {
+            case e: Exception =>
+              logError("Error while attempting tp relogin to KDC", e)
+          }
+        }
+      }
+      delegationTokenRenewer.schedule(reloginRunnable, 6, TimeUnit.HOURS)
+    }
+    val nns = getNameNodesToAccess(sparkConf)
+    obtainTokensForNamenodes(nns, conf, loggedInUGI.getCredentials)
+    val remoteFs = FileSystem.get(conf)
+    // If lastCredentialsFileSuffix is 0, then the AM is either started or restarted. If the AM
+    // was restarted, then the lastCredentialsFileSuffix might be > 0, so find the newest file
+    // and update the lastCredentialsFileSuffix.
+    if (lastCredentialsFileSuffix == 0) {
+      val credentialsPath = new Path(sparkConf.get("spark.yarn.credentials.file"))
+      listCredentialsFilesSorted(remoteFs, credentialsPath)
+        .lastOption.foreach { status =>
+        lastCredentialsFileSuffix = getSuffixForCredentialsPath(status)
+      }
+    }
+    val nextSuffix = lastCredentialsFileSuffix + 1
+    val tokenPathStr =
+      sparkConf.get("spark.yarn.credentials.file") + "-" + nextSuffix
+    val tokenPath = new Path(tokenPathStr)
+    val tempTokenPath = new Path(tokenPathStr + ".tmp")
+    logInfo("Writing out delegation tokens to " + tempTokenPath.toString)
+    val stream = Option(remoteFs.create(tempTokenPath, true))
+    try {
+      stream.foreach { s =>
+        loggedInUGI.getCredentials.writeTokenStorageToStream(s)
+        s.hflush()
+        s.close()
+        logInfo(s"Delegation Tokens written out successfully. Renaming file to $tokenPathStr")
+        remoteFs.rename(tempTokenPath, tokenPath)
```

Comment:
No, it is not because of atomic renames - but it is simply to ensure that an old credentials file does not get overwritten/renamed while being read by an executor.